game-app/game-core/src/main/java/games/strategy/engine/data/MutableProperty.java

```diff
@@ -103,6 +110,21 @@ public T getValue() {
     return getter.get();
   }
 
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  public T getValueCopy() {
+    T value = getter.get();
+    if (value instanceof List) {
+      return (T) new ArrayList((List) value);
+    } else if (value instanceof IntegerMap) {
+      return (T) new IntegerMap((IntegerMap) value);
+    } else if (value instanceof Set) {
+      return (T) new HashSet((Set) value);
+    } else if (value instanceof Map) {
+      return (T) new HashMap((Map) value);
+    }
+    return value;
+  }
```

Comment:
I know this is a bit messy, and technically not full proof.

For example, what if the getter is some other type? We don't know how to arbitrarily copy those.
Or if the contents of the collection is itself a mutable object (I've not seen this to be the case with any attachments I modified, but theoretically possible).

We don't want to do just a clone() because if the value is an unmodifiable collection, the copy will be that too.

Though... this is technically solving two distinct problems:
  1. We don't want to point to a "live" value from a Change. So we definitely need a copy.
  2. We don't want to set the actual property on the attachment to an unmodifiable collection.

I guess maybe we could handle these two separately, via:
  1. We could just call "clone()" without needing to inspect the type.
  2. Have all collection setters copy the params themselves. So we'll need to change all of those, but it's not the first time I'm editing all these properties.

@RoiEXLab WDYT?

Reply:
Actually, we can't use "clone()" because it's protected on Object and Cloneable interface doesn't define it either!

So no way to clone() without knowing the type.

Reply:
I'm 100% convinced this method should **_not_** be part of `MutableProperty` in any shape or form. This class is supposed to bundle a bunch of setters and getters together nothing else.

Regardless what do I think about this?
Honestly, this might be a legitimate use-case for reflection.
1. Get an object's class
2. check if it has a "copy-constructor" and
3. call that with the object.
4. Return the new object, or throw an exception if any of  the steps fails

Also maybe add a comment to `IntegerMap`'s copy constructor that it might be called by reflection, to prevent accidental removal (or verify this behaviour in a unit test, that works too of course)

Reply:
I don't think the reflective approach will solve the problem fully, since the problem is if the instance is an immutable collection, we want to construct a non-immutable collection.

Given we have one call site, one option is we just move this to ChangeAttachmentChange or perhaps a static on DefaultAttachment that ChangeAttachmentChange calls (e.g. copyPropertyValue() on the result of the getter).

DefaultAttachment could be a reasonable place for this given it already has helpers for dealing with these types (i.e. which is actually where the immutable views of collections come from, such as getIntegerMapProperty(), getListProperty(), etc). WDYT?

Reply:
Good point about the mutability. I really wonder though: Where is the object cast back to its original type? Could the copy be created there? Your ideas sound good nevertheless 👍

So feel free to trust your gut or play around with it for a bit. I don't think there's a definite wrong answer, just many options to weigh against each other

Reply:
For collections, the type is the interface type (i.e. List not ArrayList), since that's the type of the property. So it doesn't get cast to ArrayList, we just assign it fine to List and the problem only happens if the property is modified again and we try to add to it (since an unmodifiable List is still a List, just its mutators throw exceptions).

No casting needed on the getter side, since MutableProperty is already generic on the type.

OK, I'll go with the approach of adding a `copyPropertyValue()` to `DefaultAttachment`.